compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3903,23 +3955,28 @@ public void visit(BLangWhile whileNode, AnalyzerData data) {
                 ConditionResolver.checkConstCondition(types, symTable, whileNode.expr) == symTable.trueType
                         && !data.breakFound;
         data.breakFound = prevBreakFound;
+        analyzeOnFailClause(onFailExists, whileNode.body, whileNode.onFailClause, data);
     }
 
     @Override
     public void visit(BLangDo doNode, AnalyzerData data) {
         data.env = SymbolEnv.createTypeNarrowedEnv(doNode, data.env);
-        if (doNode.onFailClause != null) {
-            this.analyzeNode(doNode.onFailClause, data);
+        boolean onFailExists = doNode.onFailClause != null;
+        if (onFailExists) {
+            data.commonAnalyzerData.errorTypes.push(new LinkedHashSet<>());
         }
         analyzeStmt(doNode.body, data);
+        analyzeOnFailClause(onFailExists, doNode.body, doNode.onFailClause, data);
     }
 
     @Override
     public void visit(BLangFail failNode, AnalyzerData data) {
         BLangExpression errorExpression = failNode.expr;
         BType errorExpressionType = typeChecker.checkExpr(errorExpression, data.env, data.prevEnvs,
                 data.commonAnalyzerData);
-
+        if (!data.commonAnalyzerData.errorTypes.empty()) {
+            data.commonAnalyzerData.errorTypes.peek().add(types.getErrorTypes(failNode.expr.getBType()));
+        }
```

Comment:
for negative scenario, if the type of `errorExpression` is not an error, with this we are going to add type `other` to `errortypes`. Is it correct?
For example
```ballerina
import ballerina/io;

public function main() {
    int|() e = 10;
    do {
        fail e;
    } on fail var error(m) {
        io:println("Error: ", m);
    }
}


```

Reply:
This is where we add based on fail expression types, right? So I guess the error scenario will be something like

```ballerina
public function main() {
    do {
        fail e;
    } on fail var m {
        
    }
}
```

But yeah, we shouldn't add semantic error. Will fix.

Reply:
Addressed this and some other scenarios in bbbbf35.

Reply:
In the following scenario, still we add the type `other` to the `errortypes`. This happen because the `getErrorTypes` function gives the type as `other` if the type does not contain any specific `error` type. Is this correct?


```ballerina
import ballerina/io;

public function main() {
    int e = 10;
    do {
        fail e;
    } on fail var error(m) {
        io:println("Error: ", m);
    }
}

```

Reply:
Fixed in 781e3ee.

Reply:
Misunderstood earlier, thanks!